samples/nrf9160/nrf_cloud_rest_agps/src/main.c

```diff
@@ -0,0 +1,664 @@
+#include <modem/nrf_modem_lib.h>
+#include <modem/at_cmd.h>
+#include <modem/at_notif.h>
+#include <modem/modem_info.h>
+#include <net/nrf_cloud_rest.h>
+#include <modem/modem_jwt.h>
+#include <power/reboot.h>
+#include <logging/log.h>
+#include <dk_buttons_and_leds.h>
+#include <net/nrf_cloud_agps.h>
+#include <nrf_modem_gnss.h>
+#include <modem/agps.h>
+
+#if defined(CONFIG_REST_ID_SRC_COMPILE_TIME)
+BUILD_ASSERT(sizeof(CONFIG_REST_DEVICE_ID) > 1, "Device ID must be specified");
+#elif defined(CONFIG_REST_ID_SRC_IMEI)
+#define CGSN_RSP_LEN 19
+#define IMEI_LEN 15
+#define IMEI_CLIENT_ID_LEN (sizeof(CONFIG_REST_ID_PREFIX) \
+			    - 1 + IMEI_LEN)
+BUILD_ASSERT(IMEI_CLIENT_ID_LEN <= NRF_CLOUD_CLIENT_ID_MAX_LEN,
+	"REST_ID_PREFIX plus IMEI must not exceed NRF_CLOUD_CLIENT_ID_MAX_LEN");
+#endif
+
+LOG_MODULE_REGISTER(rest_agps_sample, CONFIG_NRF_CLOUD_REST_AGPS_SAMPLE_LOG_LEVEL);
+
+#define REST_RX_BUF_SZ 2100
+#define BUTTON_EVENT_BTN_NUM CONFIG_REST_BUTTON_EVT_NUM
+#define LED_NUM CONFIG_REST_LED_NUM
+#define JITP_REQ_WAIT_SEC 10
+#define NCELL_MEAS_CNT 17
+#define FAILURE_LIMIT 10
+
+static char device_id[NRF_CLOUD_CLIENT_ID_MAX_LEN + 1];
+
+static K_SEM_DEFINE(button_press_sem, 0, 1);
+static K_SEM_DEFINE(lte_connected, 0, 1);
+static K_SEM_DEFINE(cell_data_ready, 0, 1);
+static K_SEM_DEFINE(do_agps_request_sem, 0, 1);
+
+static struct modem_param_info modem_info;
+static struct lte_lc_cell current_cell;
+
+static struct nrf_modem_gnss_agps_data_frame agps_request;
+static struct nrf_modem_gnss_pvt_data_frame last_pvt;
+static struct nrf_modem_gnss_nmea_data_frame last_nmea;
+static bool jitp_requested;
+
+static bool disconnected = true;
+
+void agps_print_enable(bool enable);
+
+static int set_device_id(void)
+{
+	int err = 0;
+
+#if defined(CONFIG_REST_ID_SRC_COMPILE_TIME)
+
+	memcpy(device_id, CONFIG_REST_DEVICE_ID, strlen(CONFIG_REST_DEVICE_ID));
+	return 0;
+
+#elif defined(CONFIG_REST_ID_SRC_IMEI)
+
+	char imei_buf[CGSN_RSP_LEN + 1];
+
+	err = at_cmd_write("AT+CGSN", imei_buf, sizeof(imei_buf), NULL);
+	if (err) {
+		LOG_ERR("Failed to obtain IMEI, error: %d", err);
+		return err;
+	}
+
+	imei_buf[IMEI_LEN] = 0;
+
+	err = snprintk(device_id, sizeof(device_id), "%s%.*s",
+		       CONFIG_REST_ID_PREFIX,
+		       IMEI_LEN, imei_buf);
+	if (err < 0 || err >= sizeof(device_id)) {
+		return -EIO;
+	}
+
+	return 0;
+
+#elif defined(CONFIG_REST_ID_SRC_INTERNAL_UUID)
+
+	struct nrf_device_uuid dev_id;
+
+	err = modem_jwt_get_uuids(&dev_id, NULL);
+	if (err) {
+		LOG_ERR("Failed to get device UUID: %d", err);
+		return err;
+	}
+	memcpy(device_id, dev_id.str, sizeof(dev_id.str));
+
+	return 0;
+
+#endif
+
+	return -ENOTRECOVERABLE;
+}
+
+static void print_pvt_data(struct nrf_modem_gnss_pvt_data_frame *pvt_data)
+{
+	char buf[300];
+	size_t len;
+
+	len = snprintf(buf, sizeof(buf),
+		      "\r\n\tLongitude:  %f\r\n\t"
+		      "Latitude:   %f\r\n\t"
+		      "Altitude:   %f\r\n\t"
+		      "Speed:      %f\r\n\t"
+		      "Heading:    %f\r\n\t"
```

Comment:
You could add also accuracy and speed accuracy here. Also, would be nice to round these to one decimal and maybe add units for each value.